apps/api/src/app/services/content/tags.service.ts

```diff
@@ -0,0 +1,42 @@
+import { Inject, Injectable, NotFoundException } from "@nestjs/common";
+import { TagsStore } from "@dragonfish/api/database/content/stores";
+import { TagKind, TagsModel, TagsForm } from "@dragonfish/shared/models/content";
+import { TagsTree } from "@dragonfish/shared/models/content/tags.model";
+import { IContent, ITagsService } from "../../shared/content";
+
+@Injectable()
+export class TagsService implements ITagsService {
+    constructor(
+        private readonly tagsStore: TagsStore,
+        @Inject('IContent') private readonly contentService: IContent) { }
+
+    async fetchTags(kind: TagKind): Promise<TagsModel[]> {
+        return await this.tagsStore.fetchTags(kind);
+    }
+
+    async fetchDescendants(id: string): Promise<TagsTree> {
+        return await this.tagsStore.fetchDescendants(id);
+    }
+
+    async createTag(kind: TagKind, form: TagsForm): Promise<TagsModel> {
+        return await this.tagsStore.createTag(form, kind);
+    }
+
+    async updateTag(id: string, form: TagsForm): Promise<TagsModel> {
+        return await this.tagsStore.updateTag(id, form);
+    }
+
+    async deleteTag(id: string): Promise<void> {        
+        // First, delete all references to this tag elsewhere in the database
+        const tagToDelete = await this.tagsStore.findOne(id);
+        if (!tagToDelete) {
+            throw new NotFoundException(`No tag with the ID ${id} could be found.`);
+        }
+
+        await this.contentService.removeTagReferences(tagToDelete._id);
+
+        // Then, delete the tag itself
+        const deletedDocument = await this.tagsStore.deleteTag(id);
+        return deletedDocument;
+    }
```

Comment:
This method is a great example of why I wanted a separation between controllers and stores. Now we have a service that can pull in the Store of the thing it's in charge of, as well as a another service. The TagsStore now doesn't need to know a damn thing about Content, and the controller doesn't get tied up with business logic.

Reply:
yeah we should really look into separating out concerns more than we do currently. i've been really guilty of that